feat: openapi typescript client (#2885)
Redocly/redocly-cli: CONTRIBUTING.md

### Unit tests

Run unit tests with this command: `npm run unit`.

> This command runs the suite for every package whose tests match the discovery glob — there is no per-package `npm test` script.

Author: Fixed - pack:prepare now packs client-generator.tgz too (with its openapi-core dependency pointed at the sibling tgz, same as respect-core), and CONTRIBUTING documents installing it for runtime: package apps. The CLI tarball itself didn't need it — the published CLI bundles the generator.

Run `npm test` to start both unit and e2e tests (and additionally typecheck the code).

### Monorepo test conventions

This is a monorepo with NPM workspaces.
All tests share a single Vitest installation and a single configuration file at the repository root.
Do not add a per-package `vitest.config.ts`, a per-package `test` script, or per-package `vitest` / `@vitest/*` dependencies — they will be ignored or conflict with the root setup.

The root configuration is in [`vitest.config.ts`](./vitest.config.ts) and defines:

- The `unit` suite, which discovers tests via the glob `packages/*/src/**/*.test.ts`.
- The `e2e` suite, which discovers tests under `tests/e2e/**`.
- Coverage (Istanbul provider) collected for packages enumerated in `coverage.include`.
- Repo-wide minimum coverage thresholds, plus optional per-glob overrides for packages that want stricter limits.

Vitest globals (`describe`, `it`, `expect`, `vi`, `beforeEach`, `afterEach`, …) are enabled and the TypeScript types for them are provided through [`tsconfig.json`](./tsconfig.json)'s `"types": ["vitest/globals", "node"]`.

> Do **not** add `import { describe, it, expect } from 'vitest'` to test files — those names are already in scope.

Reviewer: ```suggestion
Do **not** add `import { describe, it, expect } from 'vitest'` to test files — these names are already in scope.
```

### Where tests live

Tests live next to the source they cover, inside a sibling `__tests__/` folder, and use the `.test.ts` suffix:

```text
packages/<your-package>/
  src/
    feature.ts
    __tests__/
      feature.test.ts        ← unit tests
    submodule/
      thing.ts
      __tests__/
        thing.test.ts
```

> The root config automatically picks them up; no additional wiring is needed for **discovery**.

Reviewer: ```suggestion
The root config picks up the test files automatically.
There is no need to add additional wiring for **discovery**.
```

### Adding tests for a new package

> When introducing a new package under `packages/`, do the following to plug it into the existing test infrastructure:

Reviewer: ```suggestion
When introducing a new package under `packages/`, plug it into the existing test infrastructure:
```

> 1. Author tests under `packages/<your-package>/src/**/__tests__/*.test.ts`. Use the Vitest globals — no imports from `'vitest'`.

Reviewer: ```suggestion
1. Author tests under `packages/<your-package>/src/**/__tests__/*.test.ts`.
    Use the Vitest globals — no imports from `'vitest'`.
```

2. Open the root [`vitest.config.ts`](./vitest.config.ts) and append your package's source glob to `coverage.include`, for example:

   ```typescript
   include: [
     'packages/cli/src/**/*.ts',
     'packages/core/src/**/*.ts',
     'packages/respect-core/src/**/*.ts',
     'packages/<your-package>/src/**/*.ts',
   ],
   ```

3. If your package contains pure type-definition modules (files that compile to empty `.js` like `types.ts` or `model.ts`), add them to `coverage.exclude` so they don't dilute the coverage signal.

> 4. Optionally enforce stricter per-file coverage for your package using a per-glob threshold alongside the repo-wide minimums:

Reviewer: ```suggestion
4. (Optional) Enforce stricter per-file coverage for your package using a per-glob threshold alongside the repo-wide minimums:
```

   ```typescript
   thresholds: {
     lines: 80,
     functions: 83,
     statements: 80,
     branches: 72,
     'packages/<your-package>/src/**/*.ts': {
       lines: 100,
       functions: 100,
       statements: 100,
       branches: 100,
     },
   },
   ```

> 5. Do not declare `vitest` or `@vitest/coverage-istanbul` in the new package's `package.json`. They are workspace-wide dev dependencies, installed once at the root.

Reviewer: ```suggestion
5. Do not declare `vitest` or `@vitest/coverage-istanbul` in the new package's `package.json`.
    These are workspace-wide dev dependencies, installed once at the root.
```

### Unit tests

Run unit tests with this command: `npm run unit`.

> This runs the suite for every package whose tests match the discovery glob — there is no per-package `npm test` script.

Reviewer: ```suggestion
This command runs the suite for every package whose tests match the discovery glob — there is no per-package `npm test` script.
```

Unit tests in the **cli** package are sensitive to top-level configuration file (**redocly.yaml**).

To run tests from a single file, run: `npm run unit -- <path/to/your/file.test.ts>`.
To run a specific test, use this command: `npm run unit -- -t 'Test name'`.
To update snapshots, run `npm run unit -- -u`.

Run `npm run unit` with coverage reporting always enabled (the `coverage` block in the root config sets `enabled: true`); the HTML report is written to `coverage/`.

### E2E tests

Run e2e tests with this command: `npm run e2e`.

E2E tests are sensitive to any additional output (like `console.log`) in the source code.

> To update snapshots, run `npm run e2e -- -u`. This includes the file-based snapshots used by some tests via `toMatchFileSnapshot` (for example, `tests/e2e/generate-client/cafe.snapshot.ts` — the committed full-file output of the TypeScript client generator). Always review snapshot diffs in the pull request to confirm the change is intentional.

Reviewer: ```suggestion
To update snapshots, run `npm run e2e -- -u`.
This command includes the file-based snapshots used by some tests via `toMatchFileSnapshot`.
For example, `tests/e2e/generate-client/cafe.snapshot.ts` is the committed full-file output of the TypeScript client generator.
Always review snapshot diffs in the pull request to confirm the change is intentional.
```

Having `redocly.yaml` in the root of the project affects the unit tests, and console logs affect the e2e tests, so make sure to get rid of both before running tests.
Run `npm test` to start both unit and e2e tests (and additionally typecheck the code).

> ### Monorepo test conventions

Reviewer 2: These additions are mostly duplications or explanations of a common testing approaches. Noone ever asked us to clarify that. I would refrain from adding this info to the contribution guide as it makes it harder to read and find actually useful info.

Author: removed

### Unit tests

Run unit tests with this command: `npm run unit`.
This command runs the suite for every package whose tests match the discovery glob — there is no per-package `npm test` script.

Unit tests in the **cli** package are sensitive to top-level configuration file (**redocly.yaml**).

To run tests from a single file, run: `npm run unit -- <path/to/your/file.test.ts>`.
To run a specific test, use this command: `npm run unit -- -t 'Test name'`.
To update snapshots, run `npm run unit -- -u`.
To skip coverage, run it with `--coverage=false`.

> Run `npm run unit` with coverage reporting always enabled (the `coverage` block in the root config sets `enabled: true`); the HTML report is written to `coverage/`.

Reviewer 2: ```suggestion
```

It's just a common configuration. And the coverage is discussed in the previous paragraph.

Author: corrected
